### ISSUE TITLE
The Fastest Google Fonts

### ISSUE THREAD

**Issue by nirjan-dev** [nirjan-dev/learning-to-code#378]
[link to the post](https://csswizardry.com/2020/05/the-fastest-google-fonts/)

# font-display: swap;

For this test, I added the &display=swap back in. In effect, this makes the font files themselves asynchronous—the browser immediately displays our fallback text before swapping to the web font whenever it arrives. This means we’re not going to leave users looking at any invisible text (FOIT), which makes for both a faster and more pleasant experience.

N.B. It’s only more pleasant if you make the effort to define a suitable fallback to display in the interim—flashing a page full of Times New Roman before settling on Open Sans would likely be a net worse experience. Thankfully, Monica has made this process not only easy, but surprisingly fun. I wouldn’t be able to do this bit without her Font Style Matcher.

```html
<link rel="stylesheet"
      href="$CSS&display=swap" />
```

# Async CSS

This will implicitly tell the browser to load the CSS file in a non-blocking fashion, applying the styles only to the print context. However, the moment the file arrives, we tell the browser to apply it to all contexts, thus styling the rest of the page.

```html
<link rel="stylesheet"
      href="$CSS&display=swap"
      media="print" onload="this.media='all'" />
```

While the trick is devilishly simple—which is what makes it so cool—I’ve long had my reservations. Y’see, a regular, synchronous stylesheet blocks rendering, so a browser will assign it Highest priority. A print stylesheet (or any stylesheet that doesn’t match the current context) is assigned the priority at the complete opposite end of the spectrum: Idle.

This means that as browsers begin making requests, the asynchronous CSS file often gets grossly under-prioritised (or rather, it’s prioritised correctly, but probably way less than you expect). Take for example Vitamix, a client for whom I implemented asynchronous CSS for their own font provider(s):

# preload
Okay, so if print CSS is too low priority, what we need is a high priority asynchronous fetch. Enter preload.

Complementing our media-trick stylesheet with a preload ensures we get the best of all worlds:

    an asynchronous high priority fetch that will work in almost all modern browsers, and;
    a very widely supported method for reapplying CSS that we loaded asynchronously.

N.B. We can’t go full preload as it isn’t widely enough supported. In fact, at the time of writing, about 20% of this site’s visitors would be unable to make use of it. Consider the print stylesheet a fallback.

```html

<link rel="preload"
      as="style"
      href="$CSS&display=swap" />

<link rel="stylesheet"
      href="$CSS&display=swap"
      media="print" onload="this.media='all'" />
```

# preconnect

Put another way: even though Google are preconnecting the fonts.gstatic.com origin for us, they’re only gaining about a 10ms head-start. Put another-other way, this file is latency-bound, not bandwidth-bound.

If we implement a first-party preconnect, we should stand to make some pretty huge gains. Let’s see what happens.

```html
<link rel="preconnect"
      href="https://fonts.gstatic.com"
      crossorigin />

<link rel="preload"
      as="style"
      href="$CSS&display=swap" />

<link rel="stylesheet"
      href="$CSS&display=swap"
      media="print" onload="this.media='all'" />
```

# Bonus: font-display: optional;

Using asynchronous CSS and font-display leaves us susceptible to FOUT (or, hopefully, FOFT if we’ve designed our Fallbacks properly). To try and mitigate this, I decided to run a test using font-display: optional;.

This variation tells the browser that web fonts are considered optional, and if we can’t get hold of the font files during our "extremely small block period" then we offer "no swap period". The practical upshot of which is that in the event that the web fonts takes too long to load, that pageview won’t utilise it at all. This helps to prevent the FOUT which will in turn lead to a more stable experience for your user—they won’t see text restyle part-way through their pageview—and a better Cumulative Layout Shift score.

However, this proved consistently troublesome when using asynchronous CSS. When the print stylesheet gets turned into an all stylesheet, the browser updates the CSSOM then applies it against the DOM. In this moment, the page is told it needs some web fonts, and the "extremely small block period" kicks in, showing a FOIT midway through the page load lifecycle. To make things worse, the browser will replace the FOIT with the same fallback it started with, so the user doesn’t even get the benefit of a new font. It basically looks like a bug.

# Findings

While self-hosting your web fonts is likely to be the overall best solution to performance and availability problems, we’re able to design some fairly resilient measures to help mitigate a lot of these issues when using Google Fonts.

A combination of asynchronously loading CSS, asynchronously loading font files, opting into FOFT, fast-fetching asynchronous CSS files, and warming up external domains makes for an experience several seconds faster than the baseline.

This is something that I strongly recommend adopting if you are a Google Fonts user.

If Google Fonts isn’t your only render-blocking resource, and if you’re violating any of the other principles for fast CSS (e.g. if you’re @importing your Google Fonts CSS file), then your mileage will vary. These optimisations are most beneficial on project where Google Fonts is posing one of your biggest performance bottlenecks.

# Google Fonts Async Snippet

There a lot of techniques combined here, but the resulting code is still slim and maintainable enough that it should’t pose a problem. The snippet doesn’t need breaking apart and can all be kept together in the <head> of your document.

Here is the optimum snippet to use for fast Google Fonts:

```html
<!--
  - 1. Preemptively warm up the fonts’ origin.
  -
  - 2. Initiate a high-priority, asynchronous fetch for the CSS file. Works in
  -    most modern browsers.
  -
  - 3. Initiate a low-priority, asynchronous fetch that gets applied to the page
  -    only after it’s arrived. Works in all browsers with JavaScript enabled.
  -
  - 4. In the unlikely event that a visitor has intentionally disabled
  -    JavaScript, fall back to the original method. The good news is that,
  -    although this is a render-blocking request, it can still make use of the
  -    preconnect which makes it marginally faster than the default.
  -->

<!-- [1] -->
<link rel="preconnect"
      href="https://fonts.gstatic.com"
      crossorigin />

<!-- [2] -->
<link rel="preload"
      as="style"
      href="$CSS&display=swap" />

<!-- [3] -->
<link rel="stylesheet"
      href="$CSS&display=swap"
      media="print" onload="this.media='all'" />

<!-- [4] -->
<noscript>
  <link rel="stylesheet"
        href="$CSS&display=swap" />
</noscript>
```
